### PR TITLE
Use map to check whether stack trace is needed

### PR DESCRIPTION
**Pull request by tedyu**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
In Server#ServeHTTP, we use list to check whether stack trace is needed for given status.

This PR utilizes map to speed up the check.

```release-note
NONE
```

### CODE REVIEW

**Review by mattjmcnaughton — APPROVED** [review]
Thanks for this pr!

/lgtm

I'm not opposed to this pr, although to be honest, I'm not sure we're going to really see meaningful performance improvements... I ran some light performance testing between the two and it was a toss up which executed faster. In other words, I'm not sure O(n) vs O(1) matters much when n is capped at 7.

That being said, it can't hurt to have positive examples of O(1) in the code base, and I think they are equally readable.

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go]: Instead of introducing a new function, why not update `StatusIsNot` to be more perfomant by taking the provided statuses and constructing a map to use when the `StacktracePred` is called?

**Review by jpbetz — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go
Comment (tedyu) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go]: StatusIsNot may be passed various sequences of statuses.
The change in this PR optimizes the case where the statuses are known at compilation time where the map is constructed once.

If we take the give statuses on the fly and construct the map, I think it is not more performant - both are of O(N) complexity.

Thanks

Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go]: > StatusIsNot may be passed various sequences of statuses.
> The change in this PR optimizes the case where the statuses are known at compilation time where the map is constructed once.
> 
> If we take the give statuses on the fly and construct the map, I think it is not more performant - both are of O(N) complexity.

What I mean is, update `StatusIsNot` to something like:

```go
func StatusIsNot(statuses ...int) StacktracePred {
        // construct map
	return func(status int) bool {
		// map lookup
	}
}
```
Then you can create the StacktracePred just once and use it, e.g.:

```
var statusesNoTracePred = httplog.StatusIsNot(..._)
```

Comment (tedyu) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go]: I have modified the code along the lines above.

**Comment by jpbetz** [comment]
Great, thanks!

/lgtm

**Comment by liggitt** [comment]
/approve

/hold
please squash before merge

**Comment by tedyu** [comment]
@liggitt 
I have squashed the commits

**Comment by tedyu** [comment]
@jpbetz 
Please give /lgtm one more time.

Thanks

**Comment by tedyu** [comment]
@jpbetz @liggitt 
All tests passed.

**Comment by tedyu** [comment]
@jpbetz @liggitt
Please review.

Thanks

**Comment by tedyu** [comment]
@liggitt 
It seems you need to cancel the hold you placed yesterday - otherwise this won't be merged.

Thanks

**Pull request merged**
